sw/device/tests/usbdev_stream_test.c

```diff
@@ -0,0 +1,753 @@
+   */
+  usbdev_stream_t streams[STREAMS_MAX];
+  /**
+   * Per-endpoint limits on the number of buffers that may be queued for
+   * transmission
+   */
+  uint8_t tx_bufs_limit[USBDEV_NUM_ENDPOINTS];
+  /**
+   * Per-endpoint counts of completed buffers queued for transmission
+   */
+  uint8_t tx_bufs_queued[USBDEV_NUM_ENDPOINTS];
+  /**
+   * Total number of completed buffers
+   */
+  uint8_t tx_queued_total;
+  /**
+   * Buffers that have been filled but cannot yet be presented for transmission
+   * TODO - perhaps absorb the buffer queuing into usb_testutils because the dif
+   * API is explicitly not robust against back-to-back sending of multiple
+   * buffers to a single endpoint, and because the read performance is reliant
+   * upon having additional buffer(s) already available for immediate
+   * presentation
+   */
+  // 12 X 24 X 4 (or 8?)( BYTES... could perhaps simplify this at some point
+  dif_usbdev_buffer_t tx_bufs[USBDEV_NUM_ENDPOINTS][MAX_TX_BUFFERS];
+};
+
+/**
+ * Configuration values for USB.
+ * TODO - dymamically construct a config descriptor appropriate to the test;
+ *        this would avoid creating unusable ports on the host and also provide
+ *        a little more testing
+ */
+static const uint8_t config_descriptors[] = {
+    USB_CFG_DSCR_HEAD(USB_CFG_DSCR_LEN + STREAMS_MAX * (USB_INTERFACE_DSCR_LEN +
+                                                        2 * USB_EP_DSCR_LEN),
+                      STREAMS_MAX),
+
+    VEND_INTERFACE_DSCR(0, 2, 0x50, 1),
+    USB_BULK_EP_DSCR(0, 1U, USBDEV_MAX_PACKET_SIZE, 0),
+    USB_BULK_EP_DSCR(1, 1U, USBDEV_MAX_PACKET_SIZE, 0),
+
+    VEND_INTERFACE_DSCR(1, 2, 0x50, 1),
+    USB_BULK_EP_DSCR(0, 2U, USBDEV_MAX_PACKET_SIZE, 0),
+    USB_BULK_EP_DSCR(1, 2U, USBDEV_MAX_PACKET_SIZE, 0),
+};
+
+/**
+ * Test descriptor
+ */
+static const uint8_t test_descriptor[] = {
+    USB_TESTUTILS_TEST_DSCR(1, NUM_STREAMS | 0xF0U, 0, 0, 0)};
+
+/**
+ * USB device context types.
+ */
+static usb_testutils_ctx_t usbdev;
+static usb_testutils_controlep_ctx_t usbdev_control;
+
+/**
+ * Pinmux handle
+ */
+static dif_pinmux_t pinmux;
+
+/**
+ * State information for streaming data test
+ */
+static usbdev_stream_test_ctx_t stream_test;
+
+/**
+ * Specify whether to perform verbose logging, for visibility
+ *   (Note that this substantially alters the timing of interactions with the
+ * DPI model and will increase the simulation time)
+ */
+static bool verbose = false;
+
+/**
+ * Send only maximal length packets?
+ * (important for performance measurements on the USB, but obviously undesirable
+ *  for testing reliability/function)
+ */
+static bool max_packets = false;
+
+/**
+ * Number of streams to be created
+ */
+static const unsigned nstreams = NUM_STREAMS;
+
+/**
+ * Diagnostic logging; expensive
+ */
+static bool log_traffic = false;
+
+// Dump a sequence of bytes as hexadecimal and ASCII for diagnostic purposes
+static void buffer_dump(const uint8_t *data, size_t n) {
```

Comment:
We already have things like `base_hexdump`, etc. This might be a useful variant that could be made available more broadly, so maybe it should be added in a similar place. In any case, this seems to be duplicating a function that you already contributed in a recent PR (that I missed), under the name `dump_bytes`.

Reply:
Okay, thanks. I'm afraid I was simply unaware of the existing base_(f)hexdump functions; I'm happy to change that and lose the duplicated effort. Re use of base_s(n)printf, it was a log time ago now, and I perhaps thought that calling base_() functions directly was a violation of modularity.

Re the dump_bytes() function in the DPI model, Bazel's sandboxing of builds prevents code sharing between the hw/dv/dpi code and the sw/device tree, alas. I still have to raise an Issue about this, because USB-defined constants have to be specified in multiple places as a result.